### PR TITLE
global stabilization on GPU implemented

### PR DESCRIPTION
**Pull request by gridley**
With this, I was able to run the full-core 2D case on my Titan X. Below is some run time data as well.

```
[  TITLE  ]  ************************************************************************************************************************
[  TITLE  ]                                                       TIMING REPORT                                                     
[  TITLE  ]  ************************************************************************************************************************
[  RESULT ]  Total Track Generation & Segmentation Time...........7.9280E+02 sec
[  RESULT ]  Total time to solution...............................7.9791E+03 sec
[  RESULT ]    Solver initialization..............................4.1678E+01 sec
[  RESULT ]    Transport Sweep....................................7.7838E+03 sec
[  RESULT ]  Integration time by segment-group-device (sweep).....5.7986E-11 sec
[  RESULT ]  Integration time by segment-group-device (total).....5.9441E-11 sec
[SEPARATOR]  ------------------------------------------------------------------------------------------------------------------------
[  RESULT ]             # tracks          # segments          # FSRs
[SEPARATOR]  ------------------------------------------------------------------------------------------------------------------------
[  RESULT ]             298040           424448024           1285684           
[SEPARATOR]  ------------------------------------------------------------------------------------------------------------------------
Transport problem solved by openMOC 	 8833.05083990097 s
```

### CODE REVIEW

File: src/accel/cuda/GPUSolver.cu [hunk 1]
```diff
@@ -176,6 +180,60 @@ protected:
 
 };
 
+/**
```
Comment (GiudGiud) [inline src/accel/cuda/GPUSolver.cu:184]: missing a blank line

File: src/accel/cuda/GPUSolver.cu [hunk 2]
```diff
@@ -176,6 +180,60 @@ protected:
...
+  else
+  {
+    // TODO, write other stabilization types.
+    // Apparently this is the only way to throw an error from a kernel at the
+    // moment... inline PTX assembly. craziness, man
+    asm("trap;");
+  }
+}
+
+/**
```
Comment (GiudGiud) [inline src/accel/cuda/GPUSolver.cu:205]: add a line break

File: src/accel/cuda/GPUSolver.cu [hunk 3]
```diff
@@ -176,6 +180,60 @@ protected:
...
+ * @brief Stabilize the current flux with a previously computed stabilizing flux
+ * @param scalar_flux the scalar flux in each FSR and energy group
+ * @param stabilizing_flux the array of stabilizing fluxes
+ */
+__global__ void stabilizeFluxOnDevice(FP_PRECISION* scalar_flux,
+                                               FP_PRECISION* stabilizing_flux)
```
Comment (GiudGiud) [inline src/accel/cuda/GPUSolver.cu]: align arguments

File: src/accel/cuda/GPUSolver.cu [hunk 4]
```diff
@@ -176,6 +180,60 @@ protected:
...
+      stabilizing_flux[tid] = multiplicative_factor * scalar_flux[tid];
+      tid += blockDim.x * gridDim.x;
+    }
+  }
+  else
+  {
+    // TODO, write other stabilization types.
+    // Apparently this is the only way to throw an error from a kernel at the
+    // moment... inline PTX assembly. craziness, man
+    asm("trap;");
```
Comment (GiudGiud) [inline src/accel/cuda/GPUSolver.cu]: stabilization type could be checked during initialization, maybe we dont need this ?

Comment (GiudGiud) [reply src/accel/cuda/GPUSolver.cu]: apparently there's an assertionError tha can be used in CUDA kernels. Would work for our case
[link](https://stackoverflow.com/questions/50755717/triggering-a-runtime-error-within-a-cuda-kernel)

**Review by GiudGiud — CHANGES_REQUESTED** [review]

File: src/accel/cuda/GPUSolver.cu (hunk 4 above)
Comment (gridley) [reply src/accel/cuda/GPUSolver.cu]: Yeah, it's definitely not needed, true. It's always checked beforehand, and I had primarily written this to ensure I'd not made a mistake.

**Comment by gridley** [comment]
OK, the requested changes should be in now.

File: src/accel/cuda/GPUSolver.cu (hunk 4 above)
Comment (GiudGiud) [reply src/accel/cuda/GPUSolver.cu]: Let's delete it then

File: src/accel/cuda/GPUSolver.cu [hunk 5]
```diff
@@ -1423,13 +1489,25 @@ void GPUSolver::storeFSRFluxes() {
...
-  log_printf(ERROR, "computeStabilizingFlux() is not yet implemented on GPU.");
+  if (!_stabilize_transport) return;
+
+  if (_stabilization_type == GLOBAL) {
+    FP_PRECISION* scalar_flux = thrust::raw_pointer_cast(&_scalar_flux[0]);
+    FP_PRECISION* stabilizing_flux = thrust::raw_pointer_cast(&_stabilizing_flux[0]);
+    computeStabilizingFluxOnDevice<<<_B, _T>>>(scalar_flux, stabilizing_flux);
+  }
+  else
+    log_printf(ERROR, "Only global stabilization works on GPUSolver now.");
```
Comment (GiudGiud) [inline src/accel/cuda/GPUSolver.cu:1496]: The default stabilization option (DIAGONAL) makes the GPUSolver run into this error. 
Let's change the default stabilization type to be GLOBAL for the GPUSolver

**Review by GiudGiud — CHANGES_REQUESTED** [review]
looks good, works well. 
Impressive results on the 2D core. 

Last few nitpicky stuff then merging

File: src/accel/cuda/GPUSolver.cu (hunk 4 above)
Comment (gridley) [reply src/accel/cuda/GPUSolver.cu]: Ah whoops. Totally meant to make that change but didn't

File: src/accel/cuda/GPUSolver.cu (hunk 5 above)
Comment (gridley) [reply src/accel/cuda/GPUSolver.cu:1496]: OK, I just put that in the GPUSolver constructor then.

**Pull request merged**